README.md
# MMM-Rennes-StarBus

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/). It tells you when the next bus arrives at your bus stop. This is for people that live in Rennes (France).

## Installation
Navigate into your MagicMirror's `modules` folder and execute git clone `https://github.com/GilMardon/MMM-Rennes-StarBus`.

## Using the module

To use this module, add the following configuration block to the modules array in the `config/config.js` file:
```js
var config = {
    modules: [
        {
            module: 'MMM-Rennes-StarBus',
            position: "bottom_right",   // This can be any of the regions.
            config: {
                // See below for configurable options
                updateInterval: 60000,
                timeFormat : 24,
                maxEntries : 5,
                lines : [
                {
                    type: 'bus', 
                    line: '0004', 
                    stop: 'Bois+Labbé',
                    destination: 'ZA+Saint-Sulpice'
                }]
            }
        }
    ]
}
```

## Configuration options

| Option           | Description
|----------------- |-----------
| `lines`    | *Required* Fill this array with information from [www.star.fr](http://www.star.fr/).<br><br> **Type:** `array[{
                    type: 'bus', 
                    line: '0004', 
                    stop: 'Bois+Labbé',
                    destination: 'ZA+Saint-Sulpice'
                }]` <br> **Default value:** `none`
| `timeFormat`     | *Optional* Use 12 or 24 hour format. <br><br> **Possible values:** `12` or `24` <br> **Default value:** `24`
| `maxEntries`     | *Optional* The maximum number of buses to display. <br><br> **Possible values:** `1` to `10` <br> **Default value:** `5`
| `updateInterval` | *Optional* How often to check for the next bus. <br><br> **Type:** `int`<br> **Default value:** `60000` milliseconds (1 minute)


CHANGELOG.md
# MMM-Rennes-StarBus Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [1.0.0] - Unreleased
First public release


node_helper.js
/* A MagicMirror module to show bus timetables and arival times of Rennes Star Bus.
 * Copyright (C) 2018 Gil Mardon
 * https://github.com/GilMardon/MMM-Rennes-StarBus
 * License: GNU General Public License */

var NodeHelper = require("node_helper");
const restling = require('restling');
const Promise = require('bluebird');
const _ = require('lodash');

module.exports = NodeHelper.create({

    // Override socketNotificationReceived method.

	/* socketNotificationReceived(notification, payload)
	 * This method is called when a socket notification arrives.
	 *
	 * argument notification string - The identifier of the noitication.
	 * argument payload mixed - The payload of the notification.
	 */
    socketNotificationReceived: function (notification, payload) {
        var self = this;

        if (notification === "CONFIG") {
            self.config = payload;
            self.getData();

            setInterval(function () {
                self.getData();
            }, self.config.updateInterval);
        } else if (notification === "GET_DATA") {
            self.getData();
        }
    },

	/*
	 * getData
	 * function example return data and show it in the module wrapper
	 * get a URL request
	 *
	 */
    getData: function () {
        let self = this;
        let returnedData = {
            nhits : 0,
            records : []
        };
        let RestArgs = {
            rejectUnauthorized: false
        };

        return Promise.each(self.config.lines, function(line) {
            let urlApi = "https://data.rennesmetropole.fr/api/records/1.0/search/?dataset=prochains-passages-des-lignes-de-bus-du-reseau-star-en-temps-reel&sort=nomcourtligne&facet=idligne&facet=nomcourtligne&facet=sens&facet=destination&facet=precision&facet=nomarret&refine.idligne=" +
            line.line +
            "&refine.nomarret=" +
            line.stop +
            "&refine.destination=" +
            line.destination;

            return restling.get(urlApi, RestArgs).then(function (response) {
                if (response.response.statusCode === 401) {
                    self.sendSocketNotification("ERROR", this.status);
                    console.log(self.name, this.status);
                } else if (response.response.statusCode != 200) {
                    console.log(self.name, "Could not load data.");
                }
                delete response.data.parameters;
                delete response.data.facet_groups;
                
                returnedData.nhits += response.data.nhits;
                returnedData.records = returnedData.records.concat(response.data.records);
                
                return Promise.resolve();
            }).then(function() {            
                let sortedData = { 
                    records : _.sortBy(returnedData.records, 'fields.arrivee'),
                    nhits : returnedData.nhits 
                };

                return self.sendSocketNotification("DATA", sortedData);
            }).catch(function (e) {
                console.log("Communications error:", e.message);
            });
        });
    }
});


MMM-Rennes-StarBus.js
/* A MagicMirror module to show bus timetables and arival times of Rennes Star Bus.
 * Copyright (C) 2018 Gil Mardon
 * https://github.com/GilMardon/MMM-Rennes-StarBus
 * License: GNU General Public License */

Module.register("MMM-Rennes-StarBus", {
    defaults: {
        timeFormat: config.timeFormat,
        maxEntries: 5,
        updateInterval: 60000,
        retryDelay: 5000
    },

    requiresVersion: "2.1.0", // Required version of MagicMirror

    start: function () {
        var self = this;
        var dataRequest = null;
        var dataNotification = null;

        this.sendSocketNotification("CONFIG", this.config);

        // Schedule update timer.
        setInterval(function () {
            self.sendSocketNotification("GET_DATA");
        }, this.config.updateInterval);

    },
    formatTimeString: function (date) {
        var m = moment(date);

        var hourSymbol = "HH";
        var periodSymbol = "";

        if (this.config.timeFormat !== 24) {
            hourSymbol = "h";
            periodSymbol = " A";
        }

        var format = hourSymbol + ":mm" + periodSymbol;

        return m.format(format);
    },
    getDom: function () {
        var self = this;

        // Create an element wrapper for display data
        var wrapper = document.createElement("table");

        // If data are not empty
        if (this.dataRequest) {
            this.dataRequest.records.forEach(function (data, i) {
                var lineWrapper = document.createElement("tr");

                var wrapperArrivee = document.createElement("td");
                wrapperArrivee.innerHTML = data.fields.nomcourtligne;
                wrapperArrivee.className = "nomcourtligne";
                wrapper.appendChild(wrapperArrivee);

                var wrapperArrivee = document.createElement("td");
                wrapperArrivee.innerHTML = self.formatTimeString(data.fields.arrivee);
                wrapperArrivee.className = "arrivee";
                wrapper.appendChild(wrapperArrivee);

                var wrapperNomArret = document.createElement("td");
                wrapperNomArret.innerHTML = data.fields.nomarret;
                wrapperNomArret.className = "nomarret";
                wrapper.appendChild(wrapperNomArret);

                var wrapperDestination = document.createElement("td");
                wrapperDestination.innerHTML = "vers";
                wrapperDestination.className = "vers";
                wrapper.appendChild(wrapperDestination);

                var wrapperDestination = document.createElement("td");
                wrapperDestination.innerHTML = data.fields.destination;
                wrapperDestination.className = "destination";
                wrapper.appendChild(wrapperDestination);

                wrapper.appendChild(lineWrapper);
            });
        }

        return wrapper;
    },

    getScripts: function () {
        return ["moment.js"];
    },

    getStyles: function () {
        return [
            "MMM-Rennes-StarBus.css",
        ];
    },

    processData: function (data) {
        var self = this;
        this.dataRequest = data;
        self.updateDom(self.config.animationSpeed);
    },

    /**
     * socketNotificationReceived from helper
     * Processes messages from node_helper
     */
    socketNotificationReceived: function (notification, message) {
        if (notification === "DATA") {
            this.processData(message);
        } else if (notification === "ERROR") {
            self.updateDom(self.config.animationSpeed);
        }
    },
});
